table/src/main/java/com/bazaarvoice/emodb/table/db/astyanax/AstyanaxTableDAO.java

```diff
@@ -1566,4 +1595,137 @@ protected static String getMillisecondPrecisionZonedDateTime(ZonedDateTime dateT
+                        tableEventDatacenter.newTableEvent(json.getTable(), new TableEvent(TableEvent.Action.DROP, json.getUuidString()), Instant.now())
+                    )
+                    .map(delta -> {
+                        Audit audit = new AuditBuilder()
+                                .setComment(String.format("Adding events for table %s", json.getTable()))
+                                .setLocalHost()
+                                .build();
+                        return new Update(_systemTableEventRegistry, tableEventDatacenter.getDataCenter(), TimeUUIDs.newUUID(), delta, audit, WriteConsistency.GLOBAL);
+                    })
+                    .forEach(updates::add);
+        });
+
+        _backingStore.updateAll(updates);
+    }
+
+    @Override
+    public void markTableEventAsComplete(String registrationId, String table, String uuid) {
+
+        TableEventDatacenter datacenter = _objectMapper.convertValue(
+                _backingStore.get(_systemTableEventRegistry, _selfDataCenter, ReadConsistency.STRONG),
+                TableEventDatacenter.class
+        );
+
+        Audit audit = new AuditBuilder()
+                .setComment(String.format("Marking event as complete for table %s and uuid %s", table, uuid))
+                .setLocalHost()
+                .build();
+
+        Delta delta = datacenter.markTableEventAsComplete(registrationId, table, uuid);
+
+        _backingStore.update(_systemTableEventRegistry, _selfDataCenter, TimeUUIDs.newUUID(),
+                delta, audit, WriteConsistency.GLOBAL);
+
+    }
+
+    private void checkTableEventsComplete(String table, String uuid) {
+        Iterator<Map<String, Object>> tableEventDatacenterIterator = _backingStore.scan(_systemTableEventRegistry, null, LimitCounter.max(), ReadConsistency.STRONG);
+
+        List<Update> updates = StreamSupport.stream(Spliterators.spliteratorUnknownSize(tableEventDatacenterIterator, 0), false)
+                .map(json -> _objectMapper.convertValue(json, TableEventDatacenter.class))
+                .map(tableEventDatacenter -> {
+                    Delta delta = tableEventDatacenter.markTableEventAsReady(table, uuid);
+                    if (delta != null) {
+                        Audit audit = new AuditBuilder()
+                                .setComment(String.format("Marking event ready for table %s", table))
+                                .setLocalHost()
+                                .build();
+                        return new Update(_systemTableEventRegistry, tableEventDatacenter.getDataCenter(), TimeUUIDs.newUUID(), delta, audit, WriteConsistency.GLOBAL);
+                    }
+
+                    return null;
+                })
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+
+
+        if (!updates.isEmpty()) {
+            _backingStore.updateAll(updates);
+            throw new IllegalStateException("Table events still pending");
```

Comment:
I was thinking about retrying the check after certain time, but like u pointed if the Maintenance Scheduler retry can do this job of rerunning the Maintenance OP, we should be good. Although proving it locally with a small retry delay might be useful.

Reply:
I already have! and i can confirm that it works!